### PR TITLE
SSE: Add Mode to drop NaN/Inf/Null in Reduction operations

### PR DESCRIPTION
**Pull request by kylebrandt**
**What this PR does / why we need it**:
Currently, query can result in a time series that can contain non-numeric values such as NaN, null, +\-Inf. In that case the Reduce expression almost always (all functions that execute math operations on values) result in NaN. This behavior basically makes Reduce operation useless and leaves only one option "Classical Condition" that can handle non-numeric values.

This PR updates Reduce expression to be more flexible in handling non-numeric values. It introduces a new setting "Mode" for Reduce expression that controls how the input series is pre-processed before applying the reduction function. The setting has 3 options:
1. Strict - Do not pre-process the input time series. This is the default behavior.
2. Drop Non-numeric Values - Filters out all non-numeric values from the input series. The reduction function is applied to the processed series. Therefore, it changes meaning of reduction functions (like Mean) as they are applied to a series with less data points.
3. Replace Non-numeric Values - Replaces all non-numeric values with some constant value. The value can be changed by the user (default 0). In this mode the reduction function is applied to the series where all non-numeric values are replaced. 

image

image

TODO:
 - [x] Docs
 - [x] Tests
 - [x] Code Cleanup
 - [x] Add option to object received from FE
 - [x] Add to FE

**Which issue(s) this PR fixes**:
Related to https://github.com/grafana/grafana/issues/40392

### CODE REVIEW

File: pkg/expr/mathexp/reduce.go [hunk 1]
```diff
@@ -80,33 +83,106 @@ func Last(fv *Float64Field) *float64 {
...
 	}
 	number.SetValue(f)
-
 	return number, nil
 }
+
+type ValueMapper interface {
+	MapSeries(s Series) Series
+	MapResult(v *float64) *float64
+}
```
Comment (kylebrandt) [inline pkg/expr/mathexp/reduce.go:136]: Seems like a good pattern generally. Perhaps:

```
type ReduceMapper interface {
    InputMapper(x *float64) *float64
    OutputMapper(x *float64) *float64 
}
```

 - "ReduceMapper" I think is more clear for its use case
 - There is already a "Result" type and the "Value" property in the code :-/, so I think intput / output is fine
 - I think both funcs can have the same signature? That way when fulfilling the interface people don't need to write the "NewSeries", "AppendPoint" loop etc., just pass the function that is handled in the body of the loop. Then that loop can move to the `Reduce` method.

Comment (yuri-tceretian) [reply pkg/expr/mathexp/reduce.go:136]: Good idea 👍🏻

**Comment by gillesdemey** [comment]
CodeQL seems to have caught a few things @yuri-tceretian are these false positives?

**Comment by yuri-tceretian** [comment]
> CodeQL seems to have caught a few things @yuri-tceretian are these false positives?

Yes, neither of files mentioned in the output were touched in this PR.

**Review by kylebrandt — COMMENTED** [review]
Changes since I started this look good to me.

Couple issues, that are likely okay to fix in the beta I think?

1. "None" should maybe be "strict" or something (nulls/nans in data generally results nans"
2.  Need to update the expressions docs markdown

For reference, here is a body for Grafana Managed Rule that uses Grafana's test datasource:

image

When using this I noticed that switching the Function, Mode, etc does not run the queries. I'm not sure if it should but that threw me for a few seconds, but perhaps it used to be that way and it was just what I am used to.

**Comment by kylebrandt** [comment]
Sofia noticed that "Last" with a null value at end panics:

```
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,20,90,30,5,null"
        },
        {
          "datasource": {
            "type": "__expr__",
            "uid": "__expr__"
          },
          "expression": "A",
          "hide": false,
          "reducer": "last",
          "refId": "B",
          "settings": {
            "mode": ""
          },
          "type": "reduce"
        }
      ],
      "title": "None-Last",
      "type": "timeseries"
    }
```

Maybe something with merging that in @yuri-tceretian ?

**Review by papagian — APPROVED** [review]
LGTM

**Comment by kylebrandt** [comment]
> I think that only need to check for `nil` [here](https://github.com/grafana/grafana/blob/de04f19c4709439717239153c7910235235a0182/pkg/expr/mathexp/reduce.go#L79) but that function was not introduced in this PR.

I think:

```go
func Last(fv *Float64Field) *float64 {
	var f float64
	if fv.Len() == 0 {
		f = math.NaN()
		return &f
	}
	v := fv.GetValue(fv.Len() - 1)
	f = *v
	return &f
}
```

to

```go
func Last(fv *Float64Field) *float64 {
	var f float64
	if fv.Len() == 0 {
		f = math.NaN()
		return &f
	}
	v := fv.GetValue(fv.Len() - 1)
	return &v
}
```

**Pull request merged**